daft/runners/pyrunner.py

```diff
@@ -210,113 +217,108 @@ def run_iter_tables(
-                                next_step.set_result(materialized_results)
-
-                            else:
-                                # Submit the task for execution.
-                                logger.debug("Submitting task for execution: %s", next_step)
-
-                                # update progress bar
-                                pbar.mark_task_start(next_step)
-
-                                future = thread_pool.submit(
-                                    self.build_partitions,
-                                    next_step.instructions,
-                                    next_step.inputs,
-                                    next_step.partial_metadatas,
-                                )
-                                # Register the inflight task and resources used.
-                                future_to_task[future] = next_step.id()
-
-                                inflight_tasks[next_step.id()] = next_step
-                                inflight_tasks_resources[next_step.id()] = next_step.resource_request
-
-                            next_step = next(plan)
-
-                    # Await at least one task and process the results.
-                    assert (
-                        len(future_to_task) > 0
-                    ), "Scheduler deadlocked! This should never happen. Please file an issue."
-                    done_set, _ = futures.wait(list(future_to_task.keys()), return_when=futures.FIRST_COMPLETED)
-                    for done_future in done_set:
-                        done_id = future_to_task.pop(done_future)
-                        del inflight_tasks_resources[done_id]
-                        done_task = inflight_tasks.pop(done_id)
-                        materialized_results = done_future.result()
-
-                        pbar.mark_task_done(done_task)
-
-                        logger.debug(
-                            "Task completed: %s -> <%s partitions>",
-                            done_id,
-                            len(materialized_results),
-                        )
-
-                        done_task.set_result(materialized_results)
+                            # Submit the task for execution.
+                            logger.debug("Submitting task for execution: %s", next_step)
+
+                            # update progress bar
+                            pbar.mark_task_start(next_step)
+
+                            future = self._thread_pool.submit(
+                                self.build_partitions,
+                                next_step.instructions,
+                                next_step.inputs,
+                                next_step.partial_metadatas,
+                            )
+                            # Register the inflight task and resources used.
+                            future_to_task[future] = next_step.id()
+
+                            self._inflight_tasks[next_step.id()] = next_step
```

Comment:
I don't know if the `id` is guaranteed to be unique across dataframe executions?
let's add an assert that  
```
next_step.id() not in self._inflight_tasks_resources
```

Reply:
Just checked: I think it should be

We have a singleton `ID_GEN = itertools.count()` that is used to generate the IDs, and it is supposedly threadsafe thanks to the GIL.

Reply:
(Also added the assert)